### PR TITLE
Add jsg::MemoryTracker docs to jsg/readme

### PR DESCRIPTION
**Pull request by jasnell**

### CODE REVIEW

File: src/workerd/jsg/README.md [hunk 1]
```diff
@@ -1551,3 +1551,99 @@ KJ_DEFER(key->reset());
 // with the scope exits.
 ```
 
+## jsg::MemoryTracker and heap snapshot detail collection
```
Comment (hoodmane) [inline src/workerd/jsg/README.md:1554]:
```suggestion
## `jsg::MemoryTracker` and heap snapshot detail collection
```

File: src/workerd/jsg/README.md [hunk 2]
```diff
@@ -1551,3 +1551,99 @@ KJ_DEFER(key->reset());
...
+and investigation of memory issues in an application (e.g. hunting down
+memory leaks, detecting bugs, optimizing memory usage, etc) so the information
+should include details that are most useful for those purposes.
+This code is only ever called when a heap snapshot is being generated so
+typically it should have very little cost. Heap snapshots are generally
+fairly expensive to create, however, so care should be taken not to make
+things too complicated. Ideally, none of the implementation methods in a
+type should allocate. There is some allocation occuring internally while
+building the graph, of course, but the methods for visitation (in particular
+the jsgGetMemoryInfo(...) method) should not perform any allocations if it
```
Comment (hoodmane) [inline src/workerd/jsg/README.md:1648]:
```suggestion
the `jsgGetMemoryInfo(...)` method) should not perform any allocations if it
```

**Review by hoodmane — APPROVED** [review]

**Comment by kentonv** [comment]
I realize it's a bit late for this but I wonder why we had to build a parallel visitation mechanism rather than reuse/extend visitForGc()?

File: src/workerd/jsg/README.md [hunk 3]
```diff
@@ -1551,3 +1551,99 @@ KJ_DEFER(key->reset());
...
+collect details for the graph. Note that this method is NOT expected to be called
+within the scope of an `IoContext`. It *will* be called while within the isolate
+lock, however.
+
+Types may also implement the following additional methods to further customize
+how they are represented in the graph:
+
+ * `v8::Local<v8::Object> jsgGetMemoryInfoWrapperObject();`
+ * `MemoryInfoDetachedState jsgGetMemoryInfoDetachedState() const;`
+ * `bool jsgGetMemoryInfoIsRootNode() const;`
```
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/jsg/README.md:1593]: Will add that separately

**Review by jclee — APPROVED** [review]

**Comment by jasnell** [comment]
> ... why we had to build a parallel visitation mechanism rather than reuse/extend visitForGc()?

We can work on combining these separately if we want. I did this as a separate path largely to avoid touching the more functionally critical `visitForGc()` (wanted to avoid accidentally introducing any bugs in that path) and because the visitation here is a bit different (tracking information on fields that typically are not tracked for gc visitation). I want to be able to experiment and tweak this independently. Once we're sure it's stable, reconciling the two into a single visitation mechanism is likely certainly possible.

**Pull request merged**